bokeh/document/json.py

```diff
@@ -62,7 +66,9 @@ class ModelChanged(TypedDict):
     model: Ref
     attr: str
     new: Unknown
-    hint: DocumentPatched | None
+
+    # mypy does not support recursive types https://github.com/python/mypy/issues/731
+    hint: DocumentPatched | None  # type: ignore
```

Comment:
I would not mind making this better in 3.0 e.g. by allowing events to adapt themselves into sub-events, rather than by taking on a "hint". But if you have any better ideas for this now, please lmk.

Reply:
Definitively hints need to go. However, I think that the type may be more specific than `DocumentPatched` (I'm not sure right now), which could avoid this limitation in mypy. Otherwise, protocols do support recursive definitions, but may not be applicable here, at least without major changes, thus it may not be worthwhile to attempt this. If `type: ignore` is going to be kept, then as in other places, this needs to be as specific statement as possible.